src/dtx/dtx_common.c

```diff
@@ -0,0 +1,632 @@
+
+		rc = vos_dtx_fetch_committable(cont->sc_hdl,
+					       DTX_THRESHOLD_COUNT, &dte);
+		if (rc <= 0)
+			break;
+
+		rc = dtx_commit(cont_hdl->sch_pool->spc_uuid, cont->sc_uuid,
+				dte, rc, cont_hdl->sch_pool->spc_map_version);
+		dtx_free_committable(dte);
+	} while (rc >= 0);
+
+	if (rc < 0)
+		D_ERROR(DF_UUID": Fail to flush CoS cache: rc = %d\n",
+			DP_UUID(cont->sc_uuid), rc);
+
+	D_ASSERT(cont->sc_dtx_flush_cbdata != NULL);
+
+	future = cont->sc_dtx_flush_cbdata;
+	rc = ABT_future_set(future, NULL);
+	D_ASSERTF(rc == ABT_SUCCESS, "ABT_future_set failed for DTX flush on "
+		  DF_UUID": rc = %d\n", DP_UUID(cont->sc_uuid), rc);
+}
+
+void
+dtx_batched_commit(void *arg)
+{
+	struct dss_module_info	*dmi = dss_get_module_info();
+
+	while (1) {
+		ABT_bool			 state;
+		struct dtx_batched_commit_args	*dbca;
+		struct ds_cont_hdl		*cont_hdl;
+		struct ds_cont			*cont;
+		struct dtx_entry		*dte = NULL;
+		struct dtx_stat			 stat = { 0 };
+		int				 rc;
+
+		if (d_list_empty(&dmi->dmi_dtx_batched_list))
+			goto check;
+
+		dbca = d_list_entry(dmi->dmi_dtx_batched_list.next,
+				    struct dtx_batched_commit_args, dbca_link);
+		cont_hdl = dbca->dbca_cont_hdl;
+		cont = cont_hdl->sch_cont;
+		if (cont->sc_closing) {
+			d_list_del(&dbca->dbca_link);
+			dtx_flush_committable(cont_hdl);
+			/* drop the ref for dtx_batched_commit_register(). */
+			ds_cont_hdl_put(cont_hdl);
+			D_FREE_PTR(dbca);
+			goto check;
+		}
+
+		d_list_move_tail(&dbca->dbca_link, &dmi->dmi_dtx_batched_list);
+		vos_dtx_stat(cont->sc_hdl, &stat);
+
+		if ((stat.dtx_committable_count > DTX_THRESHOLD_COUNT) ||
+		    (stat.dtx_oldest_committable_time != 0 &&
+		     dtx_hlc_age2sec(stat.dtx_oldest_committable_time) >
+		     DTX_COMMIT_THRESHOLD_AGE)) {
+			rc = vos_dtx_fetch_committable(cont->sc_hdl,
+						DTX_THRESHOLD_COUNT, &dte);
+			if (rc > 0) {
+				rc = dtx_commit(cont_hdl->sch_pool->spc_uuid,
+					cont->sc_uuid, dte, rc,
+					cont_hdl->sch_pool->spc_map_version);
+				dtx_free_committable(dte);
+
+				if (cont->sc_closing) {
+					d_list_del(&dbca->dbca_link);
+					dtx_flush_committable(cont_hdl);
+					ds_cont_hdl_put(cont_hdl);
+					D_FREE_PTR(dbca);
+					goto check;
+				}
+
+				if (!cont->sc_dtx_aggregating)
+					vos_dtx_stat(cont->sc_hdl, &stat);
+			}
+		}
+
+		if (!cont->sc_dtx_aggregating &&
+		    ((stat.dtx_committed_count > DTX_AGG_THRESHOLD_CNT) ||
+		     (stat.dtx_oldest_committed_time != 0 &&
+		      dtx_hlc_age2sec(stat.dtx_oldest_committed_time) >
+				DTX_AGG_THRESHOLD_AGE_UPPER))) {
+			ds_cont_get(cont);
+			cont->sc_dtx_aggregating = 1;
+			rc = dss_ult_create(dtx_aggregate, cont,
+				DSS_ULT_AGGREGATE, DSS_TGT_SELF, 0, NULL);
+			if (rc != 0) {
+				cont->sc_dtx_aggregating = 0;
+				ds_cont_put(cont);
+			}
+		}
+
+check:
+		rc = ABT_future_test(dmi->dmi_xstream->dx_shutdown, &state);
+		D_ASSERTF(rc == ABT_SUCCESS, "%d\n", rc);
+		if (state == ABT_TRUE) {
+			while (!d_list_empty(&dmi->dmi_dtx_batched_list)) {
```

Comment:
move out the inner loop below out from the big loop? So we could put the future test in an inline function as the condition of the big loop.
```
bool dtx_is_shutdown(dmi)
{
     ABT_future_test(...)
     return state == ABT_TRUE;
}

then the loop can be:

  While (!dtx_is_shutdown(dmi)) {
         ....
  }

  while (!d_list_empty(&dmi->dmi_dtx_batched_list)) {
        ....
  }
```

Reply:
OK